tripal/tripal
# Checking for new fields

#

### Issue #1793 


### Tripal Version: 4.x

## Description


This PR supports adding new fields to content types.   The support for this has been added to the `TripalFieldCollection` class which is used to provide service for adding fields to content types.  Currently, this service is used to import the fields associated to content types in the YML files when a module is installed.   This PR adds a `discover()` function to the service.  The `discover()` function gets called in the `TripalEntityUIController` when someone clicks the button "Check for new fields" on the field management page of a content type.  To find new field instances, each field can now implement it's own `discover()` method too.  It should return an array equivalent to the one we use in the YML file for adding a field to a content type.  The `TripalFieldCollection::discover()` function will call the discover function of the fields.

**What's missing:**
- There is no form for allowing a site admin to select which fields to import. This was something we spoke about and wanted to do so that if a site admin deletes a field it won't keep coming back if they don't want it to.  I felt we could add that later but just allowing users to be able to add fields for now was necessary.
- This PR is missing some functional test, so it will remain a draft until that's ready.  
- This PR only adds  a `discover()` function to the Chado properties field. Other fields that need to discover new instances of themselves will need to add a `discover()` method.

## Testing?



1. Checkout this branch
2. Create an _Oryza sativa_ organism page.
3. Create an analysis page for a genome assembly titled "MSU Rice Genome Annotation Project Release 7"
4. Download Chromosome1 of the rice GFF3 file for the rice genome here:  http://rice.uga.edu/pub/data/Eukaryotic_Projects/o_sativa/annotation_dbs/pseudomolecules/version_7.0/chr01.dir/Chr1.gff3
5. Import the GFF using the Tripal GFF3 importers
6. Publish the gene pages for rice.
7. Go to the page for managing fields for the "Gene" content type and clikc the button to "Check for new fields". You should see a new "Note" and "Score" field added.
8. Remove one or both of the new fields.  Then click the "Check for new fields" button. They should be readded.
